ci: Update Clang in "tidy" job

hebasto opened the pull request:
This PR:

1. Updates to [IWYU 0.25](https://github.com/include-what-you-use/include-what-you-use/releases/tag/0.25), which is compatible with Clang 21.

2. Fixes new "modernize-use-default-member-init" warnings. The warning in `interpreter.cpp` is a [false positive](https://github.com/llvm/llvm-project/issues/160394), so it has been suppressed.

vasild reviewed (approved):
ACK ecacfac7ead178ace3b0ec9393a7c63b5d8d3576

vasild reviewed (approved):
ACK 1bb0f7f8b6d927046fefa2eb29d0132713415cca

ryanofsky reviewed (approved):
Code review ACK 1bb0f7f8b6d927046fefa2eb29d0132713415cca. Nice code cleanups, though did leave one comment about the NOLINT.

re: https://github.com/bitcoin/bitcoin/pull/33445#issuecomment-3315082319

> Could you please take a look at the following IPC-related warning:

Thanks, created https://github.com/capnproto/capnproto/pull/2417 to address it

vasild reviewed (approved):
ACK 5b20d172ca2a46a2b525201b4ff2444f9d415d8c

ryanofsky commented on src/test/CMakeLists.txt:
```diff
@@ -177,6 +177,8 @@ if(ENABLE_IPC)
   )
   target_link_libraries(test_bitcoin bitcoin_ipc_test bitcoin_ipc)
+
+  configure_file(.clang-tidy.in .clang-tidy USE_SOURCE_PERMISSIONS COPYONLY)
```
In commit "clang-tidy: Disable `ArrayBound` check in src/ipc and src/test" (5b20d172ca2a46a2b525201b4ff2444f9d415d8c)

Kind of a shame to need this outside of the IPC directory. Maybe there should be a `src/ipc/test` subdirectory like `src/wallet/test` to keep IPC stuff grouped together.

hebasto replied:
> I was thinking of it as a followup for a future PR...

I agree with that.

> ... but I'm actually not sure how big the change would be and whether it might require code changes.

Something like a978c6bdb41fd3f8461908c30bb409a26fa62d47 in https://github.com/hebasto/bitcoin/commits/250925-llvm-tidy/.

ryanofsky reviewed (approved):
Code review ACK 5b20d172ca2a46a2b525201b4ff2444f9d415d8c. Just added link to upstream modernize-use-default-member-init bug (thanks for looking into that and reporting) and added new suppressions for capnproto clang-tidy errors since last review

fanquake commented on src/script/interpreter.cpp:
```diff
@@ -1254,8 +1253,12 @@ class CTransactionSignatureSerializer
     const CScript& scriptCode; //!< output script being consumed
     const unsigned int nIn;    //!< input index of txTo being signed
     const bool fAnyoneCanPay;  //!< whether the hashtype has the SIGHASH_ANYONECANPAY flag set
+    // Temporary workaround for a clang-tidy bug.
```
Given the upstream fix has landed (and was possibly backported?), this workaround should either be removed entirely, or updated to indicate the version of Clang which this was fixed in.

hebasto committed:
ci: Update Clang in "tidy" job

This change updates to IWYU 0.25, which is compatible with Clang 21.
Fixes new "modernize-use-default-member-init" warnings.
The warning in `interpreter.cpp` is a false positive, so it has been
suppressed.

hebasto committed:
clang-tidy: Disable `ArrayBound` check in src/ipc and src/test

hebasto commented:
Rebased and addressed [feedback](https://github.com/bitcoin/bitcoin/pull/33445#discussion_r2469984562) from @fanquake.

hebasto commented:
@vasild @ryanofsky 

Kindly asking you to refresh your ACKs after the comment [update](https://github.com/bitcoin/bitcoin/pull/33445#issuecomment-3457125965).

ryanofsky reviewed (approved):
Code review ACK 5d784bebaff5e3acc0b5180ee51d9a16aec0e356, just adding clang version comment since last review.

Would still be nice to remove ipc build code from the test directory as a followup too (https://github.com/bitcoin/bitcoin/pull/33445#discussion_r2379651340)
